campaignion_tracking/package/src/gtm.js

```diff
@@ -1,34 +1,143 @@
-// eslint-disable-next-line camelcase
-let campaignion_tracking = window.campaignion_tracking
-// eslint-disable-next-line camelcase
-if (typeof campaignion_tracking === 'undefined') {
-  console.debug('No campaignion_tracking found. Doing nothing.')
-  return
+/**
+ * Mapping of codes to eventNames.
+ *
+ * This also is a whitelist of which events should *also* be recognized as
+ * codes.
+ */
+export const codes = {
+  ds: 'donationSuccess'
 }
 
-if (campaignion_tracking.debug) {
-  console.debug('campaignion_tracking_gtm')
-}
+/**
+ * Tracker for Google Tag Manager.
+ *
+ * Implements behaviour to dispatch events to GTM.
+ */
+export class GTMTracker {
+  /**
+   * Constructor.
+   *
+   * @param {TrackerManager} tracker the shared tracker manager object
+   * @param {Boolean} debug set to true for debugging
+   */
+  constructor (tracker, debug = false) {
+    this.debug = debug
+    this.tracker = tracker
 
-let tracker = campaignion_tracking.tracker
-if (typeof tracker === 'undefined') {
-  console.debug('No TrackerManager found. Doing nothing.')
-  return
-}
+    this.printDebug('init')
 
-export const codes = {
-  cb: 'checkoutBegin',
-  ce: 'checkoutEnd',
-  ds: 'donationSuccess'
-}
+    if (typeof this.tracker === 'undefined') {
+      this.printDebug('No TrackerManager found. Doing nothing.')
+    }
+
+    /**
+     * Subscribe to fragment events.
+     *
+     * The tracker listens for specific parts in the URL fragement and
+     * published messages with event codes to trigger.
```

Comment:
Why does it need to listen for both? Isn’t the TrackerManager supposed to turn fragments and on-page events into a unified source of pubsub-messages?

Reply:
No. The TrackerManager currently should only turn fragments into code messages. It does not care what they mean, just adds an prefix. The interested trackers themselves are responsible to listen to them. So they can decide not to be interested in them.

Could be changed of course...

Reply:
move the code listener to the common module for the Drupal integration and changed the messages chnnel to just `code` (not eg. `code:gtm`) any more...